doc/dev/sharp_bits.rst

```diff
@@ -0,0 +1,995 @@
+...     return x ** 2
+>>> f(2.)
+x = Traced<ShapedArray(float64[], weak_type=True)>with<DynamicJaxprTrace(level=1/0)>
+array(4.)
+>>> f(3.)
+array(9.)
+
+We can see that on the first execution, program capture/tracing occurs, and we
+can see the dynamic variable is printed (tracers capture *type*
+and *shape*, but not numeric value). This captured program is compiled, and
+then the binary is executed directly to return the function value --- the
+print statement is never invoked with the numerical value of ``x``.
+
+When we execute the function again, steps (1) and (2) are skipped since we
+have already compiled a binary; the binary is called directly to get the function
+result, and again the print statement is never hit.
+
+This allows us to distinguish between computations that happen
+at **compile-time** (steps 1 and 2), such as the ``print`` statement above,
+and those that happen at **runtime** (step 3).
+
+.. note::
+
+    As a general rule of thumb, for a function that is repeatedly executed
+    with different parameters, we want as much evaluation as possible to
+    happen at compile time.
+
+    However, computations at compile time cannot depend on the value of
+    dynamic variable, since this is not known yet. It can only depend
+    on **static variables**, those whose values are known.
+
+.. note::
+
+    A general guideline when working with JIT compilation and Catalyst:
+
+    - Python control flow and third party libraries like NumPy and SciPy will
+      be evaluated at compile-time, and can only accept static variables.
+
+    - JAX functions, such as ``jax.numpy``, and Catalyst functions like
+      :func:`~.cond` and :func:`~.for_loop` will be evaluated at
+      runtime, and can accept dynamic variables.
+
+    Note that if AutoGraph is enabled, Catalyst will attempt to convert Python
+    control flow to its Catalyst equivalent to support dynamic variables.
+
+For example, consider the following:
+
+>>> @qjit
+... def f(x):
+...     if x > 5:
+...       x = x / 2
+...     return x ** 2
+>>> f(2.)
+TracerBoolConversionError: Attempted boolean conversion of traced array with shape bool[]..
+The error occurred while tracing the function f at <ipython-input-15-2aa7bf60efbb>:1 for make_jaxpr.
+This concrete value was not available in Python because it depends on the value of the argument x.
+See https://jax.readthedocs.io/en/latest/errors.html#jax.errors.TracerBoolConversionError
+
+This function will fail, as the Python ``if`` statement cannot accept a
+dynamic variable (a JAX tracer) as an argument.
+
+Instead, we can use Catalyst control flow :func:`~.cond` here:
+
+>>> @qjit
+... def f(x):
+...     @cond(x > 5.)
+...     def g():
+...         return x / 2
+...     @g.otherwise
+...     def h():
+...         return x
+...     return g() ** 2
+>>> f(2.)
+array(4.)
+>>> f(6.)
+array(9.)
+
+Here, both conditional branches are compiled, and only evaluated at runtime
+when the value of ``x`` is known.
```

Comment:
```suggestion
Here, both conditional branches are compiled, and only evaluated at runtime
when the value of ``x`` is known.

Note that, if the Python ``if`` statement depends only on values that are
static (known at compile time), this is fine --- the ``if`` statement will
simply be evaluated at compile time rather than runtime:

Let's consider an example where a for loop is evaluated at compile time:

>>> @qjit
... def f(x):
...     for i in range(2):
...         print(i, x)
...         x = x / 2
...     return x ** 2
>>> f(2.)
0 Traced<ShapedArray(float64[], weak_type=True)>with<DynamicJaxprTrace(level=1/0)>
1 Traced<ShapedArray(float64[], weak_type=True)>with<DynamicJaxprTrace(level=1/0)>
array(0.25)

Here, the for loop is evaluated at compile time (notice the multiple tracers
that have been printed out during program capture --- one for each loop!),
rather than runtime.
```